imap_processing/spice/kernels.py

```diff
@@ -0,0 +1,158 @@
+        * SUCCESS? Great! We're done.
+        * SpiceyError? Go to step 2.
+
+    2. Furnish metakernel at SPICE_METAKERNEL
+        * SUCCESS? Great, return the original function again (so it can be
+          re-run).
+        * KeyError? Seems like SPICE_METAKERNEL isn't set, no problem. Go to
+          step 3.
+
+    3. Did we get the parameter time_kernels_only=True?
+    --> YES? We only need LSK and SCLK kernels to run this function. Go fetch
+        those and furnish and return the original function (so it can be re-run).
+    --> NO? Dang. This is sort of the end of the line. Re-raise the error
+        generated from the failed spiceypy function call but add a better
+        message to it.
+
+    Examples
+    --------
+    There are three ways to use this object
+
+    1. A decorator with no arguments
+        >>> @ensure_spice
+        ... def my_spicey_func(a, b):
+        ...     pass
+
+    2. A decorator with parameters. This is useful
+    if we only need the latest SCLK and LSK kernels for the function involved.
+        >>> @ensure_spice(time_kernels_only=True)
+        ... def my_spicey_time_func(a, b):
+        ...     pass
+
+    3. An explicit wrapper function, providing a dynamically set value for
+    parameters, e.g. time_kernels_only
+        >>> wrapped = ensure_spice(spicey_func, time_kernels_only=True)
+        ... result = wrapped(*args, **kwargs)
+    """
+    if f_py and not callable(f_py):
+        raise ValueError(
+            f"Received a non-callable object {f_py} as the f_py argument to"
+            f"ensure_spice.  f_py must be a callable object."
+        )
+
+    def _decorator(func: Callable[..., Callable]) -> Callable:
+        """
+        Decorate or wrap input function depending on how ensure_spice is used.
+
+        Parameters
+        ----------
+        func : Callable
+            The function to be decorated/wrapped.
+
+        Returns
+        -------
+        Callable
+            If used as a function wrapper, the decorated function is returned.
+        """
+
+        @functools.wraps(func)
+        def wrapper_ensure_spice(*args: Any, **kwargs: Any) -> Any:
+            """
+            Wrap the function that ensure_spice is used on.
+
+            Parameters
+            ----------
+            *args : list
+                The positional arguments passed to the decorated function.
+            **kwargs
+                The keyword arguments passed to the decorated function.
+
+            Returns
+            -------
+            Object
+                Output from wrapped function.
+            """
+            try:
+                # Step 1.
+                return func(
+                    *args, **kwargs
+                )  # Naive first try. Maybe SPICE is already furnished.
+            except SpiceyError as spicey_err:
+                try:
+                    # Step 2.
+                    metakernel_path = os.environ["SPICE_METAKERNEL"]
```

Comment:
If kernels are already furnished, then the `Step 1` code will call the function and proceed. This allows for custom/manual furnishing of whatever kernels you want. If they are not furnished, then the secondary default behavior is to look for a metakernel pointed to by this environment variable. We can certainly tweak this functionality as we define details about how we want kernels to be furnished in the pipeline.

It seems like I should put some slides together and set up a meeting to discuss this decorator and the logic flow with the team.

Reply:
+1 I think an overview of what is happening in ppt would be helpful. 

Would we not know if kernels were already furnished? Could we use this

`with spice.KernelPool(kernels):`

Instead of this?
 `spice.furnsh(test_sclk)
    yield test_sclk
    spice.kclear()`

Reply:
We have also been putting our global configuration / env variables in the `imap-data-access` repository. I'm not sure if the SPICE configuration options should go over there too.

Reply:
+1 on an overview. That would be really helpful